packages/core/src/baseclient.ts

```diff
@@ -454,6 +523,10 @@ export abstract class BaseClient<B extends Backend, O extends Options> implement
               return;
             }
 
+            if (session) {
+              this._updateSessionFromEvent(session, finalEvent);
+            }
+
```

Comment:
I think what makes this complicated is that it is structured like this:

```js
if (shouldSkipBeforeSend) {
    // do a bunch of things *****

   return;
}

// run beforeSend

// do the same bunch of things as above *****
```

We'd avoid duplicating the several lines of code marked with `*****` if we'd reorganize it as:

```js
if (shouldRunBeforeSend) {

   // run beforeSend

}

// do a bunch of things only once
```

Reply:
Reworked from scratch.